contrib/99designs/gqlgen/tracer.go

```diff
@@ -0,0 +1,123 @@
+)
+
+const (
+	resolverObject = "resolver.object"
+	resolverField  = "resolver.field"
+	defaultResourceName = "gqlgen"
+)
+
+type gqlTracer struct {
+	cfg config
+}
+
+// New creates an a graphql.Tracer instance that can be passed to a gqlgen handler.
+// Options can be passed in for further configuration.
+func New(opts ...Option) graphql.Tracer {
+	var t gqlTracer
+	defaults(&t.cfg)
+	for _, o := range opts {
+		o(&t.cfg)
+	}
+	return &t
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) StartOperationParsing(ctx context.Context) context.Context {
+	return ctx
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) EndOperationParsing(ctx context.Context) {
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) StartOperationValidation(ctx context.Context) context.Context {
+	return ctx
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) EndOperationValidation(ctx context.Context) {
+}
+
+func (t *gqlTracer) StartOperationExecution(ctx context.Context) context.Context {
+	rctx := graphql.GetRequestContext(ctx)
+	name := defaultResourceName
+	if rctx != nil && rctx.OperationName != "" {
+		name = rctx.OperationName
+	}
+	opts := []ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeGraphql),
+		tracer.ResourceName(name),
+		tracer.ServiceName(t.cfg.serviceName),
+	}
+	if !math.IsNaN(t.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, t.cfg.analyticsRate))
+	}
+	if s, ok := tracer.SpanFromContext(ctx); ok {
+		opts = append(opts, tracer.ChildOf(s.Context()))
+	}
+	_, ctx = tracer.StartSpanFromContext(ctx, name, opts...)
+	return ctx
+}
+
+func (t *gqlTracer) StartFieldExecution(ctx context.Context, field graphql.CollectedField) context.Context {
+	span, ctx := tracer.StartSpanFromContext(ctx, "Field_"+field.Name)
+	span.SetTag("field", field.Name)
+	return ctx
+}
+
+func (t *gqlTracer) StartFieldResolverExecution(ctx context.Context, rc *graphql.ResolverContext) context.Context {
+	// This will be the span created in StartFieldExecution.
+	// StartFieldResolverExecution is called only once per StartFieldExecution, so we can add context to the
+	// span rather than starting a child span.
+	span, _ := tracer.SpanFromContext(ctx)
+	span.SetTag(ext.SpanName, rc.Object+"_"+rc.Field.Name)
+	span.SetTag(ext.ResourceName, rc.Object+"_"+rc.Field.Name)
+	span.SetTag(resolverObject, rc.Object)
+	span.SetTag(resolverField, rc.Field.Name)
+	return ctx
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) StartFieldChildExecution(ctx context.Context) context.Context {
+	return ctx
+}
+
+func (t *gqlTracer) EndFieldExecution(ctx context.Context) {
+	span, _ := tracer.SpanFromContext(ctx)
+	defer span.Finish()
+	resCtx := graphql.GetResolverContext(ctx)
+	reqCtx := graphql.GetRequestContext(ctx)
+	errList := reqCtx.GetErrors(resCtx)
+	if len(errList) != 0 {
+		span.SetTag(ext.Error, true)
```

Comment:
Does this span wrap user code too? Can this be user errors? If yes, maybe we shouldn't skip the stack.

Reply:
You're right. This is a general problem we have with the tracer. We could change it now, but it would be considered a breaking change. I'd like to suggest this plan of action:
* Only attach one single error to the span, not all of them. If there are more, simply mention them as part of the same error, just [like we do in our other GraphQL integration](https://github.com/DataDog/dd-trace-go/blob/knusbaum/gqlgen-contrib/contrib/graph-gophers/graphql-go/graphql.go#L53-L62). We should attach the error _with_ the stack trace because this is what we do most places, if not everywhere. If the stack trace turns out to be a problem, then that's a problem to be addressed as part of another issue. Currently this "workaround" is a bit of a hack and doesn't address the real issue, which might be that maybe stack traces should be off by default when setting errors.
* Open an issue under the [`v2.0` label](https://github.com/DataDog/dd-trace-go/issues?q=is%3Aopen+is%3Aissue+label%3Av2.0) and describe this problem with the stack trace as a proposal for something to address when we get to v2. When we start thinking about v2, we should include this change.
* If you think stack traces should be off by default, open a separate issue about that as well, for `v2.0`. I personally can think of a couple good arguments for it to be so.